### PR TITLE
core: `ColorTransform` cleanup

### PR DESCRIPTION
**Pull request by relrelb**
Main changes:
* Merge `ColorTransformParams` into `ColorTransformObject`, as it's only relevant for AVM1.
* Make `BitmapData::color_transform` work with a generic `ColorTransform`, which uses fixed-point arithmetic.

Note that Ruffle still calculates color transforms slightly different from Flash. This is probably caused by inaccuracy of the current `ColorTransformObject` to `ColorTransform` conversion and/or the `ColorTransform` application logic itself. Since this requires further research, it'll be fixed in a future PR.

### CODE REVIEW

File: core/src/avm1/object/color_transform_object.rs [hunk 1]
```diff
@@ -43,50 +52,42 @@ impl<'gc> ColorTransformObject<'gc> {
...
             f64
         ],
-        [set_red_offset, get_red_offset, params.red_offset, f64],
-        [set_green_offset, get_green_offset, params.green_offset, f64],
-        [set_blue_offset, get_blue_offset, params.blue_offset, f64],
-        [set_alpha_offset, get_alpha_offset, params.alpha_offset, f64],
+        [set_red_offset, get_red_offset, red_offset, f64],
+        [set_green_offset, get_green_offset, green_offset, f64],
+        [set_blue_offset, get_blue_offset, blue_offset, f64],
+        [set_alpha_offset, get_alpha_offset, alpha_offset, f64],
```
Comment (Herschel) [inline core/src/avm1/object/color_transform_object.rs:90]: While we're here, can we rename the getters to `red_multiplier` etc. to match Rust naming conventions?

File: core/src/bitmap/bitmap_data.rs [hunk 2]
```diff
@@ -480,33 +495,22 @@ impl<'gc> BitmapData<'gc> {
...
+        color_transform: ColorTransform,
     ) {
-        for x in min_x..end_x.min(self.width()) {
-            for y in min_y..end_y.min(self.height()) {
-                let color = self
-                    .get_pixel_raw(x, y)
-                    .unwrap_or_else(|| 0.into())
-                    .to_un_multiplied_alpha();
+        for x in x..(x + width).min(self.width()) {
+            for y in y..(y + height).min(self.height()) {
```
Comment (Herschel) [inline core/src/bitmap/bitmap_data.rs:505]: `saturating_add` to avoid overflow

File: core/src/avm1/globals/bitmap_data.rs [hunk 3]
```diff
@@ -596,26 +596,20 @@ pub fn color_transform<'gc>(
                 .unwrap_or(&Value::Undefined)
                 .coerce_to_object(activation);
 
-            let x = rectangle.get("x", activation)?.coerce_to_f64(activation)? as i32;
-            let y = rectangle.get("y", activation)?.coerce_to_f64(activation)? as i32;
+            let x = rectangle.get("x", activation)?.coerce_to_u32(activation)?;
+            let y = rectangle.get("y", activation)?.coerce_to_u32(activation)?;
```
Comment (Herschel) [inline core/src/avm1/globals/bitmap_data.rs:600]: The x and y don't t do wrapping behavior:
```
// Black in Flash, red square in Ruffle with this PR
import flash.display.BitmapData;
import flash.geom.Rectangle;
import flash.geom.ColorTransform;

var bData:BitmapData = new BitmapData(300, 300, false, 0);
attachBitmap(bData, 0);
var rect = new Rectangle(4294967296, 0, 200, 200);
bData.colorTransform(rect, new ColorTransform(100, 100, 100, 100, 100, 0, 0, 0));
```

File: core/src/avm1/object/color_transform_object.rs (hunk 1 above)
Comment (relrelb) [reply core/src/avm1/object/color_transform_object.rs]: Done.

File: core/src/bitmap/bitmap_data.rs (hunk 2 above)
Comment (relrelb) [reply core/src/bitmap/bitmap_data.rs]: Done.

**Comment by torokati44** [comment]
This basically reverts 57ccb714c606904af49ee9c715bc6182d442be94, which was part of #5021.
How come this is now only relevant for AVM1?
I thought the plan was to move even more `Params` objects here, for example, for all the filter effect objects.

**Comment by torokati44** [comment]
See:
Screenshot_20220910-105435

**Comment by relrelb** [comment]
> This basically reverts [57ccb71](https://github.com/ruffle-rs/ruffle/commit/57ccb714c606904af49ee9c715bc6182d442be94), which was part of #5021.
> How come this is now only relevant for AVM1?

We already have `ruffle_render::color_transform::ColorTransform` (and the equivalent `swf::ColorTransform`) as common structs. These consist of `Fixed8` and `i16` which, from my understanding, take place in the calculations.

The `f64` values are exclusive to AVM1, and this includes the various filter structs as well.
So I think we should follow the pattern of AVM1 objects that hold `f64` values, and provide a conversion to the common struct.

**Comment by torokati44** [comment]
Alright then. My original intent with these `...Params` structs was also to reduce duplication (not that I actually followed through with it then...), and you usually take good care of that, in whatever way is the most reasonable at any given time. And many things have changed since #5021, so I'm fine with going in a different direction, and yield the [dis]approval to someone else.

**Comment by relrelb** [comment]
Going to merge this in order to unblock future PRs that build upon it.

**Pull request merged**